File: python/examples/example_toml_parse.py
"""
Example: Parsing TOML from LLM output using llm_structured.

This demonstrates:
1. Basic TOML parsing
2. Parsing with JSON Schema validation
3. Extracting TOML from fenced code blocks
4. Handling arrays of tables
5. Serializing back to TOML
"""

import llm_structured as ls


def main():
    # Example 1: Simple TOML parsing
    print("=== Example 1: Simple TOML parsing ===")
    simple_toml = """
title = "TOML Example"
enabled = true
count = 42

[owner]
name = "Tom Preston-Werner"
dob = 1979-05-27T07:32:00-08:00
"""
    result = ls.loads_tomlish(simple_toml)
    print(f"Parsed: {result}")
    print()

    # Example 2: Parse and validate with JSON Schema
    print("=== Example 2: Parse and validate ===")
    config_toml = """
[server]
host = "localhost"
port = 8080

[database]
name = "myapp"
max_connections = 100
"""
    schema = {
        "type": "object",
        "properties": {
            "server": {
                "type": "object",
                "properties": {
                    "host": {"type": "string"},
                    "port": {"type": "number"}
                },
                "required": ["host", "port"]
            },
            "database": {
                "type": "object",
                "properties": {
                    "name": {"type": "string"},
                    "max_connections": {"type": "number"}
                }
            }
        },
        "required": ["server"]
    }
    validated = ls.parse_and_validate_toml(config_toml, schema)
    print(f"Valid: {validated}")
    print()

    # Example 3: Extract from fenced code block
    print("=== Example 3: Extract from markdown ===")
    llm_output = '''
Here's the configuration you requested:

```toml
[package]
name = "my-app"
version = "1.0.0"

[dependencies]
serde = "1.0"
tokio = { version = "1.0", features = ["full"] }
```

Let me know if you need any changes!
'''
    extracted = ls.extract_toml_candidate(llm_output)
    print(f"Extracted:\n{extracted}")
    parsed = ls.loads_tomlish(llm_output)
    print(f"Parsed: {parsed}")
    print()

    # Example 4: Arrays of tables
    print("=== Example 4: Arrays of tables ===")
    products_toml = """
[[products]]
name = "Hammer"
sku = 738594937

[[products]]
name = "Nail"
sku = 284758393
color = "gray"

[[products]]
name = "Screwdriver"
sku = 847382910
"""
    products = ls.loads_tomlish(products_toml)
    print(f"Products: {products}")
    for i, p in enumerate(products.get("products", [])):
        print(f"  Product {i+1}: {p['name']} (SKU: {p['sku']})")
    print()

    # Example 5: Nested tables and inline tables
    print("=== Example 5: Nested structures ===")
    nested_toml = """
[servers]
  [servers.alpha]
  ip = "10.0.0.1"
  role = "frontend"

  [servers.beta]
  ip = "10.0.0.2"
  role = "backend"

[clients]
data = [["gamma", "delta"], [1, 2]]
"""
    nested = ls.loads_tomlish(nested_toml)
    print(f"Nested: {nested}")
    print()

    # Example 6: Serialize back to TOML
    print("=== Example 6: Serialize to TOML ===")
    data = {
        "title": "My Config",
        "database": {
            "host": "localhost",
            "port": 5432,
            "enabled": True
        },
        "features": ["auth", "cache", "logging"]
    }
    toml_output = ls.dumps_toml(data)
    print(f"Serialized TOML:\n{toml_output}")

    # Example 7: Extended parsing with metadata
    print("\n=== Example 7: Extended parsing with metadata ===")
    result_ex = ls.loads_tomlish_ex(simple_toml)
    print(f"Value: {result_ex['value']}")
    print(f"Metadata: {result_ex['metadata']}")


if __name__ == "__main__":
    main()

File: typescript/examples/example_toml_parse.ts
/**
 * Example: Parsing TOML from LLM output using llm_structured.
 * 
 * This demonstrates:
 * 1. Basic TOML parsing
 * 2. Parsing with JSON Schema validation
 * 3. Extracting TOML from fenced code blocks
 * 4. Handling arrays of tables
 * 5. Serializing back to TOML
 */

import {
  loadsTomlish,
  loadsTomlishEx,
  extractTomlCandidate,
  parseAndValidateToml,
  dumpsToml,
  JsonSchema,
  JsonValue,
} from "../src/index";

function main() {
  // Example 1: Simple TOML parsing
  console.log("=== Example 1: Simple TOML parsing ===");
  const simpleToml = `
title = "TOML Example"
enabled = true
count = 42

[owner]
name = "Tom Preston-Werner"
dob = 1979-05-27T07:32:00-08:00
`;
  const result = loadsTomlish(simpleToml);
  console.log("Parsed:", JSON.stringify(result, null, 2));
  console.log();

  // Example 2: Parse and validate with JSON Schema
  console.log("=== Example 2: Parse and validate ===");
  const configToml = `
[server]
host = "localhost"
port = 8080

[database]
name = "myapp"
max_connections = 100
`;
  const schema: JsonSchema = {
    type: "object",
    properties: {
      server: {
        type: "object",
        properties: {
          host: { type: "string" },
          port: { type: "number" },
        },
        required: ["host", "port"],
      },
      database: {
        type: "object",
        properties: {
          name: { type: "string" },
          max_connections: { type: "number" },
        },
      },
    },
    required: ["server"],
  };
  const validated = parseAndValidateToml(configToml, schema);
  console.log("Valid:", JSON.stringify(validated, null, 2));
  console.log();

  // Example 3: Extract from fenced code block
  console.log("=== Example 3: Extract from markdown ===");
  const llmOutput = `
Here's the configuration you requested:

\`\`\`toml
[package]
name = "my-app"
version = "1.0.0"

[dependencies]
serde = "1.0"
tokio = { version = "1.0", features = ["full"] }
\`\`\`

Let me know if you need any changes!
`;
  const extracted = extractTomlCandidate(llmOutput);
  console.log("Extracted:", extracted);
  const parsed = loadsTomlish(llmOutput);
  console.log("Parsed:", JSON.stringify(parsed, null, 2));
  console.log();

  // Example 4: Arrays of tables
  console.log("=== Example 4: Arrays of tables ===");
  const productsToml = `
[[products]]
name = "Hammer"
sku = 738594937

[[products]]
name = "Nail"
sku = 284758393
color = "gray"

[[products]]
name = "Screwdriver"
sku = 847382910
`;
  const products = loadsTomlish(productsToml) as { products: Array<{ name: string; sku: number }> };
  console.log("Products:", JSON.stringify(products, null, 2));
  products.products?.forEach((p, i) => {
    console.log(`  Product ${i + 1}: ${p.name} (SKU: ${p.sku})`);
  });
  console.log();

  // Example 5: Serialize back to TOML
  console.log("=== Example 5: Serialize to TOML ===");
  const data = {
    title: "My Config",
    database: {
      host: "localhost",
      port: 5432,
      enabled: true,
    },
    features: ["auth", "cache", "logging"],
  };
  const tomlOutput = dumpsToml(data);
  console.log("Serialized TOML:");
  console.log(tomlOutput);

  // Example 6: Extended parsing with metadata
  console.log("\n=== Example 6: Extended parsing with metadata ===");
  const resultEx = loadsTomlishEx(simpleToml);
  console.log("Value:", JSON.stringify(resultEx.value, null, 2));
  console.log("Metadata:", JSON.stringify(resultEx.metadata, null, 2));
}

main();
